imap_processing/ena_maps/utils/corrections.py

```diff
@@ -0,0 +1,299 @@
+            Number of iterations run.
+        """
+        n_levels = len(observed_fluxes)
+        energy_levels = np.arange(n_levels) + 1
+
+        # Initial power-law estimate from observed fluxes
+        gamma_initial, _ = self.estimate_power_law_slope(observed_fluxes, energies)
+
+        # Initial source flux estimate
+        eta_initial = self.eta_esa(energy_levels, gamma_initial)
+        source_fluxes_n = observed_fluxes / eta_initial
+
+        for _iteration in range(max_iterations):
+            # Store previous iteration
+            source_fluxes_prev = source_fluxes_n.copy()
+
+            # Predictor step
+            gamma_pred, _ = self.estimate_power_law_slope(source_fluxes_n, energies)
+            gamma_half = 0.5 * (gamma_initial + gamma_pred)
+
+            # Predictor source flux estimate
+            eta_half = self.eta_esa(energy_levels, gamma_half)
+            source_fluxes_half = observed_fluxes / eta_half
+
+            # Corrector step
+            gamma_corr, _ = self.estimate_power_law_slope(source_fluxes_half, energies)
+            gamma_n = 0.5 * (gamma_pred + gamma_corr)
+
+            # Final source flux estimate for this iteration
+            eta_final = self.eta_esa(energy_levels, gamma_n)
+            source_fluxes_n = observed_fluxes / eta_final
+            source_uncertainties = observed_uncertainties / eta_final
+
+            # Check convergence
+            ratios_sq = (source_fluxes_n / source_fluxes_prev) ** 2
+            chi_n = np.sqrt(np.mean(ratios_sq)) - 1
+
+            if chi_n < convergence_threshold:
+                break
+
+        return source_fluxes_n, source_uncertainties, _iteration + 1
+
+    def apply_flux_correction(
+        self, flux: np.ndarray, flux_stat_unc: np.ndarray, energies: np.ndarray
+    ) -> tuple[np.ndarray, np.ndarray]:
+        """
+        Apply flux correction to observed fluxes.
+
+        Iterative predictor-corrector scheme is run on each spatial pixel
+        individually to correct fluxes and statistical uncertainties. This method
+        is intended to be used with the unwrapped data in the ena_maps.AbstractSkyMap
+        class or child classes.
+
+        Parameters
+        ----------
+        flux : numpy.ndarray
+            Input flux with shape (n_energy, n_spatial_pixels).
+        flux_stat_unc : np.ndarray
+            Statistical uncertainty for input fluxes. Shape must match the shape
+            of flux.
+        energies : numpy.ndarray
+            Array of energy levels in units of eV or keV.
+
+        Returns
+        -------
+        tuple[numpy.ndarray, numpy.ndarray]
+            Corrected fluxes and flux uncertainties.
+        """
+        corrected_flux = np.empty_like(flux)
+        corrected_flux_stat_unc = np.empty_like(flux_stat_unc)
+
+        # loop over spatial pixels (last dimension)
+        for i_pixel in range(flux.shape[-1]):
+            corrected_flux[:, i_pixel], corrected_flux_stat_unc[:, i_pixel], _ = (
```

Comment:
Is there any value in storing the energies returned as well? I thought we might want to look at those for what the distribution is across the map, and/or need them for other corrections later.